### ISSUE TITLE
Missing documentation on compiler specifics - grammar, AST node types, intermediate code tree Exp and Stm node types

### ISSUE THREAD

**Issue by maximusk** [Microsoft/TypeScript#17824]
I'm posting it here because it seems that this is the only way to reach out to the core team.

I'm trying to understand TypeScript compiler architecture on the low level so I'm reading the sources. [This doc page](https://github.com/Microsoft/TypeScript/wiki/Architectural-Overview#data-structures) shows that the compiler goes through the common stages of all compilers architecture: `parse tree (lexer) -> abstract syntax tree (parser) -> intermediate code -> output`. It seems that the docs on the low-level details of a compiler is currently non-existent and not sure anything of that sort will ever be officially published. 

So I'm wondering if there exists **any internal docs that can be shared** that capture the compiler architecture specifics like:

- context free and regular grammar for lexical analysis, parsing and abstract syntax
- node types for AST
- expression and statement classes specific to intermediate code tree
- etc.

**Basically anything that helps understand the construction of the typescript compiler is highly appreciated.**

**Comment by RyanCavanaugh** [comment]
We don't have any internal docs on this topic, nor time to write the quantity required to be useful. Specific questions are welcomed, though.

**Comment by mhegazy** [comment]
Also please consider our [.d.ts file](https://github.com/Microsoft/TypeScript/blob/master/lib/typescript.d.ts) as part of our documentation. the file includes definition to the AST node kinds, as well as APIs used to manipulate them.

**Comment by maximusk** [comment]
@RyanCavanaugh , @mhegazy thanks guys. I have a few specific questions:

What algorithm do you guys use for

1) lexical analysis (tokens stream)
2) syntax analysis (parse tree)

Top-down parser or bottom-up parser? Parse of ambiguous or unambiguous grammar? 

Also, I've checked the sources and it seems that the parses uses scanner to get a one token at a time instead of having a scanner to generate the stream of tokens before parsing, is it correct?

Thanks!

**Comment by mhegazy** [comment]
We use a hand-crafted recursive descent parser (so LL). the parser consumes one token at a time as you noted.

**Comment by maximusk** [comment]
@mhegazy , thanks for the info, it will be helpful when debugging the parser. Do you also use the same technique for the lexical analysis? Or do you use some finite automaton?

**Comment by DanielRosenwasser** [comment]
Everything is hand-written. The scanner is a stateful scanner which is driven by the parser when the parser is aware of the current syntactic goal. Check out: https://github.com/Microsoft/TypeScript/wiki/FAQs-for-API-Consumers#what-are-these-rescanfootoken-functions-in-the-scanner-for

> ### What are these `rescanFooToken` functions in the scanner for?

> The ECMASCript grammar describes *lexical goals* for its grammar, for which an alternate scanning rule should be used in its place from the default. These rules are triggered when ***syntactically aware consumers*** require them (i.e. ECMAScript parsers which know when a construct can occur). For details, [see the current ES spec](https://tc39.github.io/ecma262/#sec-ecmascript-language-lexical-grammar).

> One example of this is for a single `/` (the forward-slash token). As long as the `/` isn't immediately followed by another `/` (indicating a comment), the default goal (*InputElementDiv*)  is to scan it as a plain `/` or `/=` (for division operations); however, in contexts where a bare `/` or `/=` would not make sense (such as when parsing a *PrimaryExpression*), the goal is modified to *InputElementRegExp* to scan out a regular expression literal.

> The rescan functions roughly correspond to triggering the alternate rules, though rather than passing an extra parameter, a rescan is demanded of the scanner.

> Though lexical goals are not addressed in the TypeScript spec, there is effectively one new rule added for the `>`s (greater-than characters) to make generics easy to parse, but impose that any places which might encounter a `>>`, `>>>`, etc. require a rescan.

Keep in mind the scanner isn't part of our API, or if it is, it shouldn't be. 😅

**Comment by maximusk** [comment]
@mhegazy , thanks, just to clarify, you're using predictive LL(1) parser with one token lookahead that is hand-written? Or do you use recursive descent parser with recursion?

@DanielRosenwasser, thanks, what is the algorithm used for lexer? It's not finite automaton?

**Comment by gcnew** [comment]
These questions can be answered by a quick glance at the code.

If you open `scanner.ts` you'll see a function called `scan` which examines the current character (derived from the source (`text`) and current position (`pos`)) in a big, big hand-written switch statement. Where appropriate, the cases in the switch invoke helper functions.  Also, please notice the function called `speculationHelper`.

The same goes for the parser. You'll find a function called `parseSourceFile` in namespace `Parser`. From there on you can drill down the different production parsing functions. This should hint you the type of parser being used. For the second part of the question, the function `lookAhead` might be of interest to you.

**Comment by maximusk** [comment]
@gcnew , thanks for the suggestions, and I've already done that. However, given the very complex grammar of EcmaScript and my limited experience with compilers it's hard for me to recognize the pattern by reading the compiler sources, which as you've seen are huge. It's much easier to debug knowing that the parser is written, for example, as predictive recursive descent. So from the list of the top-down parser types I wish someone would just say 

> it's b, predictive recursive descent

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
